File: Reliability Analysis Tools/4. Sequential Importance Sampling/SIS_Python/SIS_aCS.py
import numpy as np
import scipy as sp
from ERANataf import ERANataf
from ERADist import ERADist
"""
---------------------------------------------------------------------------
Sequential importance sampling with adaptive conditional sampling
---------------------------------------------------------------------------
Created by:
Sebastian Geyer
Max Ehre
Iason Papaioannou
Daniel Straub

Assistant Developers:
Matthias Willer
Peter Kaplan
Luca Sardi
Daniel Koutas
Ivan Olarte Rodriguez

Engineering Risk Analysis Group
Technische Universitat Munchen
www.bgu.tum.de/era
---------------------------------------------------------------------------
Current version 2023-04
* Modification of Sensitivity Analysis Calls
---------------------------------------------------------------------------
Comments:
* The SIS-method in combination with the adaptive conditional MH sampler
  (aCS) performs well in high-dimensions. For low-dimensional problems, the
  Gaussian mixture proposal should be chosen over aCS.
* The way the initial standard deviation is computed can be changed in line 79.
  By default we use option 'a' (it is equal to one).
  In option 'b', it is computed from the seeds.
---------------------------------------------------------------------------
Input:
* N                    : number of samples per level
* p                    : N/number of chains per level
* g_fun                : limit state function
* distr                : Nataf distribution object or
                         marginal distribution object of the input variables
* burn                 : burn-in period
* tarCoV               : target coefficient of variation of the weights
* samples_return       : return of samples: 0 - none, 1 - final sample, 2 - all samples
---------------------------------------------------------------------------
Output:
* Pr       : probability of failure
* l_tot    : total number of levels
* samplesU : object with the samples in the standard normal space
* samplesX : object with the samples in the original space
* W_final  : final weights
* f_s_iid  : Independent Identically Distributed samples generated from last step
---------------------------------------------------------------------------
Based on:
1. "Sequential importance sampling for structural reliability analysis"
   Papaioannou et al.
   Structural Safety 62 (2016) 66-75
---------------------------------------------------------------------------
"""

def SIS_aCS(N, p, g_fun, distr, burn, tarCoV, samples_return):
    if (N*p != np.fix(N*p)) or (1/p != np.fix(1/p)):
        raise RuntimeError('N*p and 1/p must be positive integers. Adjust N and p accordingly')

    # initial check if there exists a Nataf object
    if isinstance(distr, ERANataf):   # use Nataf transform (dependence)
        dim = len(distr.Marginals)     # number of random variables (dimension)
        u2x = lambda u: distr.U2X(u)   # from u to x

    elif isinstance(distr[0], ERADist):   # use distribution information for the transformation (independence)
        # Here we are assuming that all the parameters have the same distribution !!!
        # Adjust accordingly otherwise or create an ERANataf object
        dim = len(distr)                    # number of random variables (dimension)
        u2x = lambda u: distr[0].icdf(sp.stats.norm.cdf(u))   # from u to x
    else:
        raise RuntimeError('Incorrect distribution. Please create an ERADist/Nataf object!')

    # LSF in standard space
    g = lambda u: g_fun(np.array(u2x(u),ndmin=2))

    # Initialization of variables and storage
    max_it   = 100              # estimated number of iterations
    m        = 0                # counter for number of levels
    samplesU = []               # space for samples in U-space

    # Properties of SIS
    nsamlev  = N                    # number of samples
    nchain   = int(nsamlev*p)       # number Markov chains
    lenchain = int(nsamlev/nchain)  # number of samples per Markov chain

    # parameters for adaptive MCMC
    opc        = 'a'   # way to estimate the std for the aCS (see comments)
    adapflag   = 1
    adapchains = int(np.ceil(100*nchain/nsamlev))
    lam        = 0.6

    # initialize samples
    accrate = np.zeros([max_it])          # space for acceptance rate
    Sk      = np.ones([max_it])           # space for expected weights
    sigmak  = np.zeros([max_it])          # space for sigmak

    # Step 1
    # Perform the first Monte Carlo simulation
    uk = sp.stats.norm.rvs(size=(nsamlev,dim))    # initial samples
    
    gk = g(uk)                                  # evaluations of g

    # save samples
    samplesU.append(uk)

    # set initial subset and failure level
    gmu = np.mean(gk)

    # Iteration
    for m in range(max_it):
        # Step 2 and 3: compute sigma and weights
        if m == 0:
            func        = lambda x: abs(np.std(sp.stats.norm.cdf(-gk/x))/ \
                                        np.mean(sp.stats.norm.cdf(-gk/x))-tarCoV)
            sigma2      = sp.optimize.fminbound(func, 0, 10.0*gmu)
            sigmak[m+1] = sigma2
            wk          = sp.stats.norm.cdf(-gk/sigmak[m+1])
        else:
            func        = lambda x: abs(np.std(sp.stats.norm.cdf(-gk/x)/sp.stats.norm.cdf(-gk/sigmak[m]))/ \
                                        np.mean(sp.stats.norm.cdf(-gk/x)/sp.stats.norm.cdf(-gk/sigmak[m]))-tarCoV)
            sigma2      = sp.optimize.fminbound(func, 0, sigmak[m])
            sigmak[m+1] = sigma2
            wk          = sp.stats.norm.cdf(-gk/sigmak[m+1])/sp.stats.norm.cdf(-gk/sigmak[m])

        # Step 4: compute estimate of expected w
        Sk[m] = np.mean(wk)
        # Exit algorithm if no convergence is achieved
        if Sk[m] == 0:
            break
        wnork = wk/Sk[m]/nsamlev    # compute normalized weights

        # Step 5: resample
        # seeds for chains
        ind = np.random.choice(np.arange(nsamlev),nchain,True,wnork)
        gk0 = gk[ind]
        uk0 = uk[ind,:]

        # Step 6: perform aCS
        # compute the standard deviation
        if opc == 'a': # 1a. sigma = ones(n,1)
            sigmaf = 1
        elif opc == 'b': # 1b. sigma = sigma_hat (sample standard deviations)
            muf    = np.mean(np.repmat(wnork,dim,1)*uk,1)
            sigmaf = np.zeros((1,dim))
            for k in range(nsamlev):
                sigmaf = sigmaf + wnork[k]*(uk[k,:]-muf)**2
        else:
            raise RuntimeError('Choose a or b')

        # compute parameter rho
        sigmafk = min(lam*sigmaf, 1)
        rhok    = np.sqrt(1-sigmafk**2)
        counta  = 0
        count   = 0

        # initialize chain acceptance rate
        alphak = np.zeros([nchain])         # initialize chain acceptance rate
        gk     = np.zeros([nsamlev])        # delete previous samples
        uk     = np.zeros([nsamlev,dim])    # delete previous samples
        for k in range(nchain):
            # set seed for chain
            u0 = uk0[k,:]
            g0 = gk0[k]
            for i in range(lenchain+burn):
                if i == burn:
                    count = count-burn

                # get candidate sample from conditional normal distribution
                ucand = np.random.normal(loc=rhok*u0, scale=sigmafk)
                #ucand = np.random.multivariate_normal(u0*lam, np.identity(dim) * (1 - lam ** 2))

                # Evaluate limit-state function
                gcand = g(ucand)

                # compute acceptance probability
                alpha     = min(1,sp.stats.norm.cdf(-gcand/sigmak[m+1])/sp.stats.norm.cdf(-g0/sigmak[m+1]))
                alphak[k] = alphak[k] + alpha/(lenchain+burn)

                # check if sample is accepted
                uhelp = sp.stats.uniform.rvs()
                if uhelp <= alpha:
                    uk[count,:] = ucand
                    gk[count]   = gcand
                    u0 = ucand
                    g0 = gcand
                else:
                    uk[count,:] = u0
                    gk[count]   = g0

                count += 1

            # adapt the chain correlation
            if adapflag == 1:
                #check whether to adapt now
                if (k+1) % adapchains == 0:

                    # mean acceptance rate of last adap_chains
                    alpha_mu = np.mean(alphak[k-adapchains+1:k+1])
                    counta   = counta+1
                    gamma    = counta**(-0.5)
                    lam      = min(np.exp(np.log(lam)+gamma*(alpha_mu-0.44)), 1)

                    # compute parameter rho
                    sigmafk = min(lam*sigmaf,1)
                    rhok    = np.sqrt(1-sigmafk**2)

        uk = uk[:nsamlev,:]
        gk = gk[:nsamlev]

        # samples return - all / all by default
        if samples_return not in [0, 1]:
            samplesU.append(uk)

        # compute mean acceptance rate of all chains in level m
        accrate[m] = np.mean(alphak)

        if sigmak[m+1] == 0:
            COV_Sl = np.nan
        else:
            COV_Sl = np.std( (gk < 0)/sp.stats.norm.cdf(-gk/sigmak[m+1]))/ \
                     np.mean((gk < 0)/sp.stats.norm.cdf(-gk/sigmak[m+1]))

        print('\nCOV_Sl =', COV_Sl)
        print('\t*aCS sigma =', sigmafk, '\t*aCS accrate =', accrate[m])
        if COV_Sl < tarCoV:
            # samples return - last
            if samples_return == 1:
                samplesU[0] = uk
            break

    # Samples return - all by default message
    if samples_return not in [0, 1, 2]:
        print("\n-Invalid input for samples return, all samples are returned by default")

    # required steps
    l_tot = m + 2

    # Calculation of the Probability of failure
    const = np.prod(Sk)
    I_final = gk <= 0
    W_final = 1 / sp.stats.norm.cdf(-gk / sigmak[m + 1])
    Pr = const * np.mean(I_final * W_final)

    # transform the samples to the physical/original space
    if samples_return != 0:
        samplesX = [u2x(samplesU[i][:,:]) for i in range(len(samplesU))]

    # resample 10000 failure samples with final weights W
    weight_id = np.random.choice(list(np.nonzero(I_final))[0], 10000, list(W_final[I_final]))
    f_s_iid = samplesX[-1][weight_id, :]

    if samples_return == 0:
        samplesU = list()  # empty return samples U
        samplesX = list()  # and X

    # Convergence is not achieved message
    if m == max_it:
        print("\n-Exit with no convergence at max iterations \n")

    return Pr, l_tot, samplesU, samplesX, W_final, f_s_iid

File: Reliability Analysis Tools/4. Sequential Importance Sampling/SIS_Matlab/SIS_aCS.m
function [Pr, l_tot, samplesU, samplesX,W_final, f_s] = ...
    SIS_aCS(N, p, g_fun, distr, burn, tarCoV, samples_return)
%% Sequential importance sampling using adaptive conditional sampling
%{
---------------------------------------------------------------------------
Created by:
Iason Papaioannou (iason.papaioannou@tum.de)
Max Ehre (max.ehre@tum.de)
Daniel Koutas
Engineering Risk Analysis Group
Technische Universitat Munchen
www.bgu.tum.de/era
---------------------------------------------------------------------------
First version: 2018-05
---------------------------------------------------------------------------
Current version: 2023-07
* Modification in the Sensitivity Analysis Output
---------------------------------------------------------------------------
Comments:
* The SIS method in combination with the adaptive conditional M-H sampler
  (aCS) performs well in high-dimensions. For low-dimensional problems, the
  Gaussian mixture proposal should be chosen over aCS.
* The way the initial standard deviation is computed can be changed in line 76.
  By default we use option 'a' (it is equal to one).
  In option 'b', it is computed from the seeds.
---------------------------------------------------------------------------
Input:
* N                    : number of samples per level
* p                    : N/number of chains per level
* g_fun                : limit state function
* distr                : Nataf distribution object or
                         marginal distribution object of the input variables
* burn                 : burn-in period
* tarCoV               : target coefficient of variation of the weights
* sensitivity_analysis : implementation of sensitivity analysis: 1 - perform, 0 - not perform
* samples_return       : return of samples: 0 - none, 1 - final sample, 2 - all samples
---------------------------------------------------------------------------
Output:
* Pr       : probability of failure
* l_tot    : total number of levels
* samplesU : object with the samples in the standard normal space
* samplesX : object with the samples in the original space
* W_final  : final weights
* f_s      : i.i.d failure samples
---------------------------------------------------------------------------
Based on:
1."Sequential importance sampling for structural reliability analysis"
   Papaioannou et al.
   Structural Safety 62 (2016) 66-75
---------------------------------------------------------------------------
%}
if (N*p ~= fix(N*p)) || (1/p ~= fix(1/p))
   error('N*p and 1/p must be positive integers. Adjust N and p accordingly');
end

%% transform to the standard Gaussian space
if any(strcmp('Marginals',fieldnames(distr))) == 1   % use Nataf transform (dependence)
   dim = length(distr.Marginals);    % number of random variables (dimension)
   u2x = @(u) reshape(distr.U2X(u),[],dim);          % from u to x
   
else   % use distribution information for the transformation (independence)
   % Here we are assuming that all the parameters have the same distribution !!!
   % Adjust accordingly otherwise or use an ERANataf object
   dim = length(distr);                    % number of random variables (dimension)
   u2x = @(u) reshape(distr(1).icdf(normcdf(u)),[],dim);   % from u to x   
end

%% LSF in standard space
g = @(u) g_fun(u2x(u)); 

%% Initialization of variables and storage
max_it = 100;    % estimated number of iterations
m      = 0;      % counter for number of levels

% Properties of SIS
nsamlev  = N;                % number of samples
nchain   = nsamlev*p;        % number Markov chains
lenchain = nsamlev/nchain;   % number of samples per Markov chain

% initialize samples
gk      = zeros(1,nsamlev);   % space for evaluations of g
accrate = zeros(max_it,1);    % space for acceptance rate
sigmak  = zeros(max_it,1);    % space for the standard deviation
Sk      = ones(max_it,1);     % space for expected weights
opc     = 'a';                % way to estimate the std for the aCS (see comments)

% parameters for adaptive MCMC
adapflag   = 1;
adapchains = ceil(100*nchain/nsamlev);  % number of chains after which the proposal is adapted
lambda     = 0.6;

samplesU  = cell(1,1);                  % space for the samples in the standard space

%% SIS aCS
%===== Step 1: Perform the first Monte Carlo simulation
uk = randn(nsamlev,dim);    % initial samples
for k = 1:nsamlev
   gk(k) = g(uk(k,:));
end

% save samples
samplesU{m+1} = uk;

% set initial subset and failure level
gmu = mean(gk);

for m = 1:max_it
   %===== Step 2 and 3: compute sigma and weights
   if m == 1
      sigma2      = fminbnd(@(x) abs( std( normcdf(-gk/x))/...
                                     mean(normcdf(-gk/x)) - tarCoV), 0, 10*gmu);
      sigmak(m+1) = sigma2;
      wk          = normcdf(-gk/sigmak(m+1));
   else
      sigma2      = fminbnd(@(x) abs( std( normcdf(-gk/x)./normcdf(-gk/sigmak(m)))/...
                                     mean(normcdf(-gk/x)./normcdf(-gk/sigmak(m))) - tarCoV ), 0, sigmak(m));
      sigmak(m+1) = sigma2;
      wk          = normcdf(-gk/sigmak(m+1))./normcdf(-gk/sigmak(m));
   end
   
   %===== Step 4: compute estimate of expected w
   Sk(m) = mean(wk);
   % exit algorithm if no convergence is achieved
   if Sk(m) == 0
      break;
   end
   wnork = wk./Sk(m)/nsamlev;          % compute normalized weights
   
   %===== Step 5: resample
   % seeds for chains
   ind = randsample(nsamlev,nchain,true,wnork);
   gk0 = gk(ind);
   uk0 = uk(ind,:);
   
   %===== Step 6: perform aCS
   % compute initial standard deviation
   switch opc
      case 'a'   % 1a. sigma = ones(n,1);
         sigmaf = 1;
      case 'b'   % 1b. sigma = sigma_hat; (sample standard deviations)
         muf    = mean(repmat(wnork,dim,1)'.*uk,1);
         sigmaf = zeros(1,dim);
         for k = 1:nsamlev
            sigmaf = sigmaf + wnork(k)*(uk(k,:)-muf).^2;
         end
      otherwise
         error('Choose a or b');
   end

   % compute parameter rho
   sigmafk = min(lambda*sigmaf, 1);
   rhok    = sqrt(1-sigmafk.^2);
   counta  = 0;
   count   = 0;
   
   % initialize chain acceptance rate
   alphak = zeros(nchain,1);
   gk     = [];                % delete previous samples
   uk     = [];                % delete previous samples
   for k = 1:nchain
      % set seed for chain
      u0 = uk0(k,:);
      g0 = gk0(k);
      for j = 1:lenchain+burn
         count = count+1;
         if j == burn+1
            count = count-burn;
         end
         % get candidate sample from conditional normal distribution
         ucand = normrnd(rhok*u0', sqrt(1-rhok^2))';
         
         % Evaluate limit-state function
         gcand = g(ucand);
         
         % compute acceptance probability
         alpha     = min(1,normcdf(-gcand/sigmak(m+1))/normcdf(-g0/sigmak(m+1)));
         alphak(k) = alphak(k)+alpha/(lenchain+burn);
         
         % check if sample is accepted
         uhelp = rand;
         if uhelp <= alpha
            uk(count,:) = ucand;
            gk(count) = gcand;
            u0  = ucand;
            g0  = gcand;
         else
            uk(count,:) = u0;
            gk(count) = g0;
         end
      end
      
      % adapt the chain correlation
      if adapflag == 1
         % check whether to adapt now
         if mod(k,adapchains) == 0
            % mean acceptance rate of last adap_chains
            alpha_mu = mean(alphak(k-adapchains+1:k));
            counta   = counta+1;
            gamma    = counta^(-0.5);
            lambda   = exp(log(lambda)+gamma*(alpha_mu-0.44));
            
            % compute parameter rho
            sigmafk = min(lambda*sigmaf,1);
            rhok    = sqrt(1-sigmafk.^2);
         end
      end
   end
   uk = uk(1:nsamlev,:);
   gk = gk(1:nsamlev);
   
   % Samples return - all / all by default
   if ~ismember(samples_return, [0 1])
       samplesU{m+1} = uk;
   end
   
   % compute mean acceptance rate of all chains in level m
   accrate(m) = mean(alphak);
   COV_Sl     = std( (gk < 0)./normcdf(-gk/sigmak(m+1)))/...
                mean((gk < 0)./normcdf(-gk/sigmak(m+1)));
   
   % it is not a problem if COV_Sl is NaN
   if ~isnan(COV_Sl)
      fprintf('\nCOV_Sl = %d\n', COV_Sl);
      fprintf('\t*aCS sigma = %g \t *aCS accrate = %g\n', sigmafk, accrate(m));
   end   
   if COV_Sl < tarCoV
      % Samples return - last
      if (samples_return == 1) 
          samplesU{1} = uk;
      end
      break;
   end
end

% Samples return - all by default message
if ~ismember(samples_return, [0 1 2])
    fprintf('\n-Invalid input for samples return, all samples are returned by default \n');
end

l_tot = m+1;

%% probability of failure
const = prod(Sk);
I_final = gk < 0;
W_final = 1./normcdf(-gk/sigmak(m+1));
Pr = const * mean(I_final .* W_final);

%% transform the samples to the physical/original space
samplesX = cell(length(samplesU),1);
if (samples_return ~= 0) 
    for i = 1:length(samplesU)
       samplesX{i} = u2x(samplesU{i});
    end
end

%% Sample Return Handling

% resample 1e4 failure samples with final weights W
weight_id = randsample(find(I_final),1e4,'true',W_final(I_final));
f_s = samplesX{end}(weight_id,:);


if samples_return == 0
    samplesU = cell(1,1);  % empty return samples U
    samplesX = cell(1,1);  % and X
end

% Convergence is not achieved message
if m == max_it
    fprintf('-Exit with no convergence at max iterations \n\n');
end

return;